ntoskrnl/ke/i386/cpu.c

```diff
@@ -1270,59 +1274,259 @@ KiCoprocessorError(VOID)
+    *((PVOID *) Save) = FsContext;
     return STATUS_SUCCESS;
 }
 
-/*
- * @implemented
+/**
+ * @brief
+ * Restores the original FPU state context that has
+ * been saved by a API call of KeSaveFloatingPointState.
+ * Callers are expected to restore the floating point
+ * state by calling this function when they've finished
+ * doing FPU operations.
+ *
+ * @param[in] Save
+ * The saved floating point context that is to be given
+ * to the function to restore the FPU state.
+ *
+ * @return
+ * Returns STATUS_SUCCESS indicating the function
+ * has fully completed its operations.
  */
+#if defined(__clang__)
+__attribute__((__target__("sse")))
+#endif
 NTSTATUS
 NTAPI
-KeRestoreFloatingPointState(IN PKFLOATING_SAVE Save)
+KeRestoreFloatingPointState(
+    _In_ PKFLOATING_SAVE Save)
 {
-    PFNSAVE_FORMAT FpState = *((PVOID *) Save);
-    ASSERT(KeGetCurrentThread()->Header.NpxIrql == KeGetCurrentIrql());
-    UNIMPLEMENTED_ONCE;
+    PFLOATING_SAVE_CONTEXT FsContext;
 
-#ifdef __GNUC__
-    asm volatile("fnclex\n\t");
-    asm volatile("frstor %0\n\t" : "=m" (*FpState));
-#else
-    __asm
+    /* Sanity checks */
+    ASSERT(Save);
+    ASSERT(KeGetCurrentIrql() <= DISPATCH_LEVEL);
+    ASSERT(KeI386NpxPresent);
+
+    /* Cache the saved FS context */
+    FsContext = *((PVOID *) Save);
+
+    /*
+     * We have to restore the regular saved FPU
+     * state. For this we must first do some
+     * validation checks so that we are sure
+     * ourselves the state context is saved
+     * properly. Check if we are in the same
+     * calling thread.
+     */
+    if (FsContext->CurrentThread != KeGetCurrentThread())
     {
-        fnclex
-        mov eax, [FpState]
-        frstor [eax]
-    };
-#endif
+        /*
+         * This isn't the thread that saved the
+         * FPU state context, crash the system!
+         */
+        KeBugCheckEx(INVALID_FLOATING_POINT_STATE,
+                     0x2,
+                     (ULONG_PTR)FsContext->CurrentThread,
+                     (ULONG_PTR)KeGetCurrentThread(),
+                     0);
+    }
+
+    /* Are we under the same NPX interrupt level? */
+    if (FsContext->CurrentThread->Header.NpxIrql != KeGetCurrentIrql())
+    {
+        /* The interrupt level has changed, crash the system! */
+        KeBugCheckEx(INVALID_FLOATING_POINT_STATE,
+                     0x1,
+                     (ULONG_PTR)FsContext->CurrentThread->Header.NpxIrql,
+                     (ULONG_PTR)KeGetCurrentIrql(),
+                     0);
+    }
+
+    /* Disable interrupts */
+    _disable();
+
+    /*
+     * The saved FPU state context is valid,
+     * it's time to restore the state. First,
+     * clear FPU exceptions now.
+     */
+    Ke386ClearFpExceptions();
+
+    /* Restore the state */
+    Ke386RestoreFpuState(FsContext->PfxSaveArea);
+
+    /* Restore the previous MXCsr if we saved it before */
+    if (KeI386XMMIPresent && KeI386FxsrPresent)
+    {
+        _mm_setcsr(FsContext->MXCsr);
+    }
```

Comment:
As far as I know, all this machinery can be replaced with `Ke386RestoreFpuState(FsContext->PfxSaveArea);`
Both MXCSR and x87 state are being overwritten by FNSAVE/FXSAVE instructions